internal/testutil/incrementaltestutil/readablebuildinfo.go

```diff
@@ -174,6 +182,27 @@ type readableBuildInfoEmitSignature struct {
 	Original            *incremental.BuildInfoEmitSignature `json:"original,omitzero"`
 }
 
+type readableBuildInfoResolvedRoot struct {
+	Resolved string
+	Root     string
+}
+
+func (b *readableBuildInfoResolvedRoot) MarshalJSON() ([]byte, error) {
+	return json.Marshal([2]string{b.Resolved, b.Root})
+}
+
+func (b *readableBuildInfoResolvedRoot) UnmarshalJSON(data []byte) error {
+	var resolvedAndRoot *[2]string
+	if err := json.Unmarshal(data, &resolvedAndRoot); err == nil {
+		*b = readableBuildInfoResolvedRoot{
+			Resolved: resolvedAndRoot[0],
+			Root:     resolvedAndRoot[1],
+		}
+		return nil
+	}
+	return fmt.Errorf("invalid BuildInfoResolvedRoot: %s", data)
```

Comment:
Nit, but typically this code is written the opposite way around, with the error check first.